File: typescript/kit/matcher/parse_segments.test.ts
import { describe, expect, it } from "vitest";
import { parseSegments } from "./parse_segments.ts";

describe("ParseSegments", () => {
	const tests = [
		{ name: "empty path", path: "", expected: [] },
		{ name: "root path", path: "/", expected: [""] },
		{ name: "simple path", path: "/users", expected: ["users"] },
		{
			name: "multi-segment path",
			path: "/api/v1/users",
			expected: ["api", "v1", "users"],
		},
		{ name: "trailing slash", path: "/users/", expected: ["users", ""] },
		{
			name: "path with parameters",
			path: "/users/:id/posts",
			expected: ["users", ":id", "posts"],
		},
		{
			name: "path with parameters, implicit index segment",
			path: "/users/:id/posts/",
			expected: ["users", ":id", "posts", ""],
		},
		{
			name: "path with parameters, explicit index segment",
			path: "/users/:id/posts/_index",
			expected: ["users", ":id", "posts", "_index"],
		},
		{ name: "path with splat", path: "/files/*", expected: ["files", "*"] },
		{
			name: "multiple slashes",
			path: "//api///users",
			expected: ["api", "users"],
		},
		{
			name: "complex path",
			path: "/api/v1/users/:user_id/posts/:post_id/comments",
			expected: [
				"api",
				"v1",
				"users",
				":user_id",
				"posts",
				":post_id",
				"comments",
			],
		},
		{
			name: "unicode path",
			path: "/café/über/resumé",
			expected: ["café", "über", "resumé"],
		},
	];

	for (const tt of tests) {
		it(tt.name, () => {
			const result = parseSegments(tt.path);
			expect(result).toEqual(tt.expected);
		});
	}
});

File: internal/matcher_tests/parse_segments_test.go
package main

import (
	"reflect"
	"runtime"
	"testing"

	"github.com/vormadev/vorma/kit/matcher"
)

func TestParseSegments(t *testing.T) {
	cases := []struct {
		name     string
		path     string
		expected []string
	}{
		{"empty path", "", []string{}},
		{"root path", "/", []string{""}},
		{"simple path", "/users", []string{"users"}},
		{"multi-segment path", "/api/v1/users", []string{"api", "v1", "users"}},
		{"trailing slash", "/users/", []string{"users", ""}},
		{
			"path with parameters",
			"/users/:id/posts",
			[]string{"users", ":id", "posts"},
		},
		{
			"path with parameters, implicit index segment",
			"/users/:id/posts/",
			[]string{"users", ":id", "posts", ""},
		},
		{
			"path with parameters, explicit index segment",
			"/users/:id/posts/_index",
			[]string{"users", ":id", "posts", "_index"},
		},
		{"path with splat", "/files/*", []string{"files", "*"}},
		{"multiple slashes", "//api///users", []string{"api", "users"}},
		{
			"complex path",
			"/api/v1/users/:user_id/posts/:post_id/comments",
			[]string{
				"api",
				"v1",
				"users",
				":user_id",
				"posts",
				":post_id",
				"comments",
			},
		},
		{
			"unicode path",
			"/café/über/resumé",
			[]string{"café", "über", "resumé"},
		},
	}

	run_matcher_backends(t, func(t *testing.T, backend matcher_test_backend) {
		for _, tc := range cases {
			t.Run(tc.name, func(t *testing.T) {
				result := backend.parse_segments(t, tc.path)
				if !reflect.DeepEqual(result, tc.expected) {
					t.Errorf(
						"ParseSegments(%q) = %v, want %v",
						tc.path,
						result,
						tc.expected,
					)
				}
			})
		}
	})
}

func BenchmarkParseSegments(b *testing.B) {
	paths := []string{
		"/",
		"/api/v1/users",
		"/api/v1/users/123/posts/456/comments",
		"/files/documents/reports/quarterly/q3-2023.pdf",
	}

	b.Run("ParseSegments", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			segments := matcher.ParseSegments(paths[i%len(paths)])
			runtime.KeepAlive(segments)
		}
	})
}
